agent/bench-scripts/postprocess/BenchPostprocess.pm

```diff
@@ -297,256 +176,297 @@ sub get_cpubusy_series {
-				$stat1_timestamp_ms = shift(@stat1_timestamps) || return;
-			}
-			my $stat1_value_base = get_value($add_from_ref, $prev_stat1_timestamp_ms);
-			# if the stat2 timestamp is different from the first $stat1 timestamp, then adjust the value based on the difference of time and values
-			my $stat2_prev_stat1_timestamp_diff_ms = ($stat2_timestamp_ms - $prev_stat1_timestamp_ms);
-			my $value_adj = 0;
-			if ($stat2_prev_stat1_timestamp_diff_ms != 0) {
-				my $stat1_prev_stat1_timestamp_diff_ms = ($stat1_timestamp_ms - $prev_stat1_timestamp_ms);
-				my $value_diff = get_value($add_from_ref, $stat1_timestamp_ms) - $stat1_value_base;
-				$value_adj = $value_diff * $stat2_prev_stat1_timestamp_diff_ms/$stat1_prev_stat1_timestamp_diff_ms;
-			}
-			my $stat1_value_interp = $stat1_value_base + $value_adj;
-			put_value($add_to_ref, $stat2_timestamp_ms,  get_value($add_to_ref,$stat2_timestamp_ms) + $stat1_value_interp);
-			$count++;
+	while ($sum_timestamps[-1] > $add_timestamps[-1]) {
+		delete($add_to_ref->{$sum_timestamps[-1]});
+		pop(@sum_timestamps);
+		if (scalar @sum_timestamps == 0) {
+			return;
 		}
 	}
-}
-sub value_exists {
-	my $params = shift;
-	my $key = $params;
-	$params = shift;
-	my $value = $params;
-	$params = shift;
-	my $array_ref = $params;
-	my $i;
-	for ($i=0; $i < scalar @{$array_ref}; $i++) {
-		if ($$array_ref[$i]{$key} == $value) {
-			return $i;
+
+	# For each timestamp in the sum, find a pair of timestamps in the addend
+	# which bracket it (where the second timestamp from the addend is greater
+	# than the timestamp from the sum and, therefore the first timestamp is
+	# less than or equal to the timestamp from the sum, since they are unique
+	# and in sorted order); perform a linear interpolation to produce a value
+	# from the addend corresponding to the timestamp from the sum and add it
+	# to the sum.  If the difference between the timestamp from the sum and
+	# the first timestamp from the addend is zero, then the interpolation,
+	# which would produce the same result, is skipped.
+	my $add_ts_1 = shift @add_timestamps;
+	my $add_ts_2 = shift @add_timestamps;
+	foreach my $sum_ts (@sum_timestamps) {
+		if ($add_ts_1 > $sum_ts) {
+			next;
+		}
+		while (defined($add_ts_2) and $add_ts_2 <= $sum_ts) {
+			$add_ts_1 = $add_ts_2;
+			$add_ts_2 = shift @add_timestamps;
+		}
+		if (!defined($add_ts_2) and $add_ts_1 != $sum_ts) {
+			# We've hit the end of the addend and the first timestamp
+			# is not an exact match, so we're done.
+			return;
 		}
+		if ($add_ts_1 > $sum_ts) {
+			die "Logic bomb:  addend timestamp (${add_ts_1}) > sum timestamp (${sum_ts})";
+		}
+		my $value = $add_from_ref->{$add_ts_1}{'value'};
+		my $time_diff = $sum_ts - $add_ts_1;
+		if ($time_diff != 0) {
+			my $val_dif = $add_from_ref->{$add_ts_2}{'value'} - $value;
+			$value += $val_dif * $time_diff/($add_ts_2 - $add_ts_1);
+		}
+		$add_to_ref->{$sum_ts}{'value'} += $value;
 	}
-	return -1;
+	return;
```

Comment:
The two instances of the code appear to be consistent with each other (I presume they were cut-and-pasted from each other originally, and my changes to them didn't disrupt the original structure); however, the algorithm is very stateful, and the difference between the two instances is inside the loop.  I _could_ implement the abstraction with a callback function parameter which could be used to invoke the custom operation, but even that is a little tricky (and I'd have to figure out how to do it properly in Perl).

So, unless you insist, I'm going to decline.  (Someday, this will all be rewritten or removed, and we can address it then, hopefully with several other significant improvements.)